ros_ws/src/interpreter_pkg/interpreter_pkg/interpreter.py

```diff
@@ -0,0 +1,168 @@
+        self.chat_publisher_=self.create_publisher(String, 'clarifying_question',10)
+        #define subscriptions
+        self.subscriber_=self.create_subscription(String, 'user_request',self.get_request,10)
+
+    def get_request(self, request_msg: String): #callback for subscribing to user response
+        #this callback gets the user response to a chat string
+        #check if the response is yes-that means user is confirming your last item,
+        #if so, the last ai message is the target
+        request=request_msg.data
+
+        if request.strip().lower()=="yes" and self.new_request==False:
+            #user is responding to a direct suggestion
+            self.target=extract_request(self.refiner.last_suggestion) 
+            self.broadcast_target(self.target)
+
+        self.new_request=False         
+
+        #SO if not responding yes, use judge to see if the item is actionable in response
+        judgement=judge(self.judger, request)
+        print("data directroy is :"+data_dir)
+        print("judge context is: "+self.judger.context)
+        print("judgement= "+ judgement)
+        if judgement=="yes":
+            #identify simple target
+            self.target=extract_request(self.extractor, request)
+            print(self.target)
+            #publish target
+            self.broadcast_target(self.target)
+
+        #if we've been going round, restart conversation
+        elif self.attempts_to_understand==5:
+            #reset rolling context
+            self.refiner.reset_context()
+            self.attempts_to_understand=0
+            #Chat the request to start over
+            start_over_msg=String()
+            start_over_msg.data="I'm sorry i dont inderstand.  Could we start over?"
+            self.chat_publisher_.publish(start_over_msg)
+        else:
+            #we dont know what the target is yet so ask another question
+            self.ask_clarifying_question(request)
+    
+    def ask_clarifying_question(self, user_msg:str):
+        self.rolling_context.append(HumanMessage(content=user_msg))
+        #LLM call to generate clarifying qustion
+        ai_question_to_user=self.refiner.chat(self.rolling_context)
+        self.last_suggestion=ai_question_to_user.content
+        self.get_logger().info( "The response from the LLM is "+ai_question_to_user.content)
+        self.rolling_context.append(ai_question_to_user.content)
+        self.attempts_to_understand+=1
+        #publish the question to user
+        clarifying_question=String()
+        clarifying_question.data=ai_question_to_user.content
+        self.chat_publisher_.publish(clarifying_question)
+
+    def broadcast_target(self, target:str): #callback for publishing target name
+        self.get_logger().info("The target is "+ self.target)
+        target_msg=String()
+        target_msg.data=target
+        self.target_publisher_.publish(target_msg)
+        self.reset()
+
+    def reset(self):
```

Comment:
Since the Node is asynchronous we may have parallel write issues depending on how you use this. You may want a read/write lock and some of these variables for that reason.

In fact, this generally doesn't work well for multiple inbound requests and only assumes a synchronous interaction. That *may* be fine but opens you up to a lot of bugs. If this thing randomly crashes due to accessing variables, this is why.

Reply:
I love to talk through and example situation of this.